source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,392 @@
+      {":method", &Http2HeaderValidator::validateMethodHeader},
+      {":authority", &Http2HeaderValidator::validateAuthorityHeader},
+      {"host", &Http2HeaderValidator::validateAuthorityHeader},
+      {":scheme", &Http2HeaderValidator::validateSchemeHeader},
+      {":path", &Http2HeaderValidator::validateGenericPathHeader},
+      {"te", &Http2HeaderValidator::validateTEHeader},
+      {"content-length", &Http2HeaderValidator::validateContentLengthHeader},
+  };
+
+  const auto& key_string_view = key.getStringView();
+  if (key_string_view.empty()) {
+    // reject empty header names
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  auto validator_it = kHeaderValidatorMap.find(key_string_view);
+  if (validator_it != kHeaderValidatorMap.end()) {
+    const auto& validator = validator_it->second;
+    return (*this.*validator)(value);
+  }
+
+  if (key_string_view.at(0) != ':') {
+    // Validate the (non-pseudo) header name
+    auto name_result = validateGenericHeaderName(key);
+    if (!name_result) {
+      return name_result;
+    }
+  } else {
+    // kHeaderValidatorMap contains every known pseudo header. If the header name starts with ":"
+    // and we don't have a validator registered in the map, then the header name is an unknown
+    // pseudo header.
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidPseudoHeader};
+  }
+
+  return validateGenericHeaderValue(value);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateResponseHeaderEntry(const HeaderString& key,
+                                                  const HeaderString& value) {
+  const auto& key_string_view = key.getStringView();
+  if (key_string_view.empty()) {
+    // reject empty header names
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  if (key_string_view == ":status") {
+    // Validate the :status header against the RFC valid range (100 <= status < 600)
+    return validateStatusHeader(StatusPseudoHeaderValidationMode::ValueRange, value);
+  } else if (key_string_view == "content-length") {
+    // Validate the Content-Length header
+    return validateContentLengthHeader(value);
+  } else if (key_string_view.at(0) != ':') {
+    auto name_result = validateGenericHeaderName(key);
+    if (!name_result) {
+      return name_result;
+    }
+  } else {
+    // The only valid pseudo header for responses is :status. If the header name starts with ":"
+    // and it's not ":status", then the header name is an unknown pseudo header.
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidPseudoHeader};
+  }
+
+  // Validate the header value
+  return validateGenericHeaderValue(value);
+}
+
+::Envoy::Http::HeaderValidator::RequestHeaderMapValidationResult
+Http2HeaderValidator::validateRequestHeaderMap(::Envoy::Http::RequestHeaderMap& header_map) {
```

Comment:
Is there a way to reject repeated pseudo-headers?

Reply:
I do not think we well see repeated pseudo headers in the header map, only the concatenated value. And it should fail character set validation if it does not contain comma. We can add tests for this.

Reply:
Yes, let's add a TODO. It will require adding more functionality to allow building raw H/2 frames, since client codec will not send a request with multiple pseudo headers.